src/main/kotlin/org/rust/lang/core/resolve/scope/RustResolveScope.kt

```diff
@@ -1,9 +1,256 @@
 package org.rust.lang.core.resolve.scope
 
-import org.rust.lang.core.psi.RustCompositeElement
-import org.rust.lang.core.psi.RustNamedElement
+import com.intellij.openapi.module.Module
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import com.intellij.psi.util.PsiTreeUtil
+import com.intellij.psi.util.PsiUtilCore
+import org.rust.cargo.util.AutoInjectedCrates
+import org.rust.cargo.util.cargoProject
+import org.rust.cargo.util.getPsiFor
+import org.rust.cargo.util.preludeModule
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.impl.RustFile
+import org.rust.lang.core.psi.impl.mixin.isStarImport
+import org.rust.lang.core.psi.impl.rustMod
+import org.rust.lang.core.psi.util.module
 
 interface RustResolveScope : RustCompositeElement {
-    val declarations: Collection<RustNamedElement>
+    data class Context(
+        val place: RustCompositeElement?,
+        val inPrelude: Boolean = false,
+        val visitedStarImports: Set<RustUseItemElement> = emptySet()
+    )
+
+    interface Entry {
+        val name: String
+        val element: RustNamedElement?
+    }
+}
+
+fun RustResolveScope.declarations(context: RustResolveScope.Context): Sequence<RustResolveScope.Entry> {
+    val visitor = RustScopeVisitor(context)
+    accept(visitor)
+    return visitor.result
+}
+
+private class ScopeEntryImpl private constructor(
```

Comment:
This split allows to make construction of `ScopeEntries` private to this file. If we move `ScopeEntry` to the `RRE` as per you suggestion then this split will indeed become redundant as we'd have to expose constructors anyway.